packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -821,8 +847,12 @@ abstract class TestRunner {
     await this.syncEngine.disableNetwork();
   }
 
-  private async doDrainQueue(): Promise<void> {
+  private async doDrainQueue(drainQueue: SpecDrainQueue): Promise<void> {
     await this.queue.drain();
+    if (drainQueue.expectUserCallback) {
```

Comment:
Hrm...  It seems like we have an artificial testing limitation to only have a single mutation-result outstanding when we drain the queue?  So we can't test the primary ack'ing two secondary writes in a row?  Does this need to be a count or something?

On that note, I'm a little nervous with how decoupled the outstandingWrites and outstandingCallbacks are and how we're not really matching them up in any way or validating that callbacks that shouldn't be called aren't getting called, etc.  Seems like there's potential for accidental bugs / oversights.  Probably not something to solve right this moment, but if you have ideas, I think it might be worth a bit of investment at some point (e.g. assigning numbers to writes / callbacks and making the callback update a "lastUserCallback" number which we can then validate against the write or something).

Reply:
Done and done :)